code/datums/components/stun_mitigation.dm

```diff
@@ -0,0 +1,183 @@
+	var/slot
+	if(parent == holder_mob.l_hand)
+		slot = SLOT_L_HAND
+	else if(parent == holder_mob.r_hand)
+		slot = SLOT_R_HAND
+	else if(parent == holder_mob.wear_mask)
+		slot = SLOT_WEAR_MASK
+	else if(iscarbon(holder_mob))
+		var/mob/living/carbon/holder_carbon = holder_mob
+		if(parent == holder_carbon.handcuffed)
+			slot = SLOT_HANDCUFFED
+		else if(parent == holder_carbon.back)
+			slot = SLOT_BACK
+		else if(ishuman(holder_mob))
+			var/mob/living/carbon/human/holder_human = holder_mob
+			if(parent == holder_human.wear_suit)
+				slot = SLOT_WEAR_SUIT
+			else if(parent == holder_human.w_uniform)
+				slot = SLOT_W_UNIFORM
+			else if(parent == holder_human.shoes)
+				slot = SLOT_SHOES
+			else if(parent == holder_human.belt)
+				slot = SLOT_BELT
+			else if(parent == holder_human.gloves)
+				slot = SLOT_GLOVES
+			else if(parent == holder_human.glasses)
+				slot = SLOT_GLASSES
+			else if(parent == holder_human.head)
+				slot = SLOT_HEAD
+			else if(parent == holder_human.wear_ear)
+				slot = SLOT_EARS
+			else if(parent == holder_human.wear_id)
+				slot = SLOT_WEAR_ID
+			else if(parent == holder_human.r_store)
+				slot = SLOT_R_STORE
+			else if(parent == holder_human.l_store)
+				slot = SLOT_L_STORE
+			else if(parent == holder_human.s_store)
+				slot = SLOT_S_STORE
+	shield_equipped(parent, holder_mob, slot)
+
+///Toggles the mitigation on or off when already equipped
+/datum/component/stun_mitigation/proc/toggle_shield/(datum/source, new_state)
+	SIGNAL_HANDLER
+	if(active == new_state)
+		return
+
+	active = new_state
+
+	if(!affected)
+		return
+
+	if(active)
+		activate_with_user()
+	else
+		deactivate_with_user()
+
+///Signal handler for equipping the shield to a slot
+/datum/component/stun_mitigation/proc/shield_equipped(datum/source, mob/living/user, slot)
+	SIGNAL_HANDLER
+	if(!(slot_flags & slot))
+		shield_detach_from_user()
+		return
+	shield_affect_user(user)
+
+///Signal handler for dropping the shield
+/datum/component/stun_mitigation/proc/shield_dropped(datum/source, mob/user)
+	SIGNAL_HANDLER
+	shield_detach_from_user()
+
+///Handles the shield setting up for a user, and activating if applicable
+/datum/component/stun_mitigation/proc/shield_affect_user(mob/living/user)
+	if(affected)
+		if(affected == user)
+			return //Already active
+		shield_detach_from_user() //changing affected
+	affected = user
+	if(active)
+		activate_with_user()
+
+///Actually activates the mitigation effect
+/datum/component/stun_mitigation/proc/activate_with_user()
+	RegisterSignal(affected, COMSIG_LIVING_PROJECTILE_STUN, PROC_REF(on_attack_stun_mitigation))
+
+///Actually deactivates the mitigation effect
+/datum/component/stun_mitigation/proc/deactivate_with_user()
+	UnregisterSignal(affected, COMSIG_LIVING_PROJECTILE_STUN)
+
+///Handles removing the mitigation from a user
+/datum/component/stun_mitigation/proc/shield_detach_from_user()
+	if(!affected)
+		return
+	SEND_SIGNAL(affected, COMSIG_MOB_SHIELD_DETATCH)
```

Comment:
de2atch
I know it's an existing signal but it should be easy enough to find/replace.